src/notes.py

```
@@ -8,49 +8,120 @@
+        self.tags = tags if tags else []
+        self.contacts = contacts if contacts else set()
 
     def edit(self, new_body: str) -> None:
-        """Edit the note by changing the body."""
+        """Edit the note by adding something to the body."""
+        self.body = self.body + " " + new_body
+
+    def replace(self, new_body: str) -> None:
+        """Edit the note by replacing the body."""
+        self.body = new_body
 
     def attach_to_contact(self, contact_name: str) -> None:
         """Attach the note to a contact."""
+        self.contacts.add(contact_name)
 
     def add_tag(self, tag: str) -> None:
         """Add a tag to the note."""
+        if len(tag) < 0 and len(tag) > 20:
+            raise ValueError("Tag must be between 1 and 20 characters")
+        self.tags.append(tag)
 
     def remove_tag(self, tag: str) -> None:
         """Remove a tag from the note."""
+        if tag not in self.tags:
+            raise ValueError(f"Tag {tag} not found")
+        self.tags.remove(tag)
+
+    def __repr__(self):
+        return f"{self.title}, {self.body}, {self.tags}, {self.contacts}"
 
 
 class NoteBook(UserDict):
     """Class representing a collection of notes."""
 
-    def add(self, title: str, body: str, tags: Optional[List[str]] = None) -> None:
+    def add(
+        self, title: str, body: str, tags: Optional[List[str]] = None, contacts: Optional[List[str]] = None
+    ) -> None:
         """Add a new note to the notebook."""
+        self.data[title] = Note(title, body, tags, contacts)
 
     def delete(self, title: str) -> None:
         """Delete a note from the notebook by title."""
+        if not title in self.data:
+            raise KeyError(f"Note with title {title} not found")
+        del self.data[title]
 
     def edit(self, title: str, new_body: str) -> None:
         """Edit the body of an existing note."""
+        if not title in self.data:
+            return KeyError(f"Note with title {title} not found")
```

need to raise error instead of return

---

HelperError